rpki-validator/src/main/java/net/ripe/rpki/validator3/util/HappyEyeballsResolver.java

```diff
@@ -273,52 +270,55 @@ private static Runnable dnsLookupRunnable(Name hostname, int queryType, Queue<Op
 
-            private boolean openConnectionFromQueue(Queue<Optional<InetAddress>> queue) throws InterruptedException {
-                final Optional<InetAddress> addressOptional = queue.poll();
-                if (addressOptional != null) {
-                    if (addressOptional.isPresent()) {
-                        final InetAddress address = addressOptional.get();
-                        try {
-                            log.debug("Connecting to {}", address);
-                            final SocketChannel socketChannel = SocketChannel.open();
-                            socketChannel.configureBlocking(false);
-                            socketChannel.connect(new InetSocketAddress(address, port));
-                            socketChannels.add(Optional.of(socketChannel));
-                        } catch (IOException e) {
-                            // can't connect - move on
-                            log.debug("Error connecting to {}", address);
-                        } finally {
-                            TimeUnit.MILLISECONDS.sleep(CONNECTION_ATTEMPT_DELAY_MS);
-                        }
-                    } else {
-                        return false;
-                    }
+            final boolean haveMoreAF2 = openConnectionFromQueue(secondAfAddresses, port, selector);
+            connectedSocket = awaitSuccessfulConnection(selector, CONNECTION_ATTEMPT_DELAY_MILLIS);
+            if (connectedSocket != null) return connectedSocket;
+
+            // watch out! do not inline these booleans,
+            // otherwise you'll shortcircuit second evaluation
+            keepGoing = (haveMoreAF1 || haveMoreAF2) && haveTime(deadline);
+        }
+        if (selector.keys().isEmpty()) {
+            return Optional.empty();
+        } else {
+            final long remainingTime = Math.max(0, deadline - System.nanoTime());
+            return awaitSuccessfulConnection(selector, TimeUnit.NANOSECONDS.toMillis(remainingTime));
+        }
+    }
+
+    private static boolean openConnectionFromQueue(Queue<Optional<InetAddress>> queue,
+                                                   int port,
+                                                   Selector selector) {
+        final Optional<InetAddress> addressOptional = queue.poll();
+        if (addressOptional != null) {
+            if (addressOptional.isPresent()) {
+                final InetAddress address = addressOptional.get();
+                try {
+                    log.info("Connecting to {}", address);
+                    final SocketChannel socketChannel = SocketChannel.open();
+                    socketChannel.configureBlocking(false);
+                    socketChannel.register(selector, SelectionKey.OP_CONNECT);
+                    socketChannel.connect(new InetSocketAddress(address, port));
+                } catch (IOException e) {
+                    // can't connect - move on
+                    log.info("Error connecting to {}", address);
```

Comment:
log.debug?

Reply:
Hm, there's whole lot of conditions which would raise an exception on programming error - like connecting on closed socket, on already connected socket, etc. So I think it needs more exposure than DEBUG.